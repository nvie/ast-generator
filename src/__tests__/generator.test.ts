import { describe, expect, it } from "vitest";

import {
  parseGrammarFromString_withClassic,
  parseGrammarFromString_withOhm,
} from "../generator";

describe("same impl", () => {
  it("simple example", () => {
    const newSyntax = `
      Abc { name: string }
    `;

    const oldSyntax = `Abc:
  name  \`string\`
`;

    expect(parseGrammarFromString_withOhm(newSyntax)).toEqual(
      parseGrammarFromString_withClassic(oldSyntax),
    );
  });

  it("same impl", () => {
    const newSyntax = `# This is a comment
   Abc {
     x0:
     string
     x1: string+
     x2: string*
     x3?:string y0:@Xyz
     y1: @Xyz*
     y2: @Xyz+
     y3?: @Xyz
     y4?: @Xyz
     z0: Pqr
     z1: Pqr+
     z2: Pqr*
     z3?: Pqr
   }
  
   @Xyz = Pqr | Stu
  
   Pqr {
     p?: number
   }
  
   # This is a comment
   Stu {
     s: number
   }
   `;

    const oldSyntax = `# This is a comment
   Abc:
     x0 \`string\`
     x1 \`string\`+
     x2 \`string\`*
     x3 \`string\`?
     y0 @Xyz
     y1 @Xyz*
     y2 @Xyz+
     y3 @Xyz?
     y4 @Xyz?
     z0 Pqr
     z1 Pqr+
     z2 Pqr*
     z3 Pqr?
  
   @Xyz:
     | Pqr
     | Stu
  
   Pqr:
     p \`number\`?
  
   # This is a comment
   Stu:
     s \`number\`
   `;

    expect(parseGrammarFromString_withOhm(newSyntax)).toEqual(
      parseGrammarFromString_withClassic(oldSyntax),
    );
  });
});

describe("parsing grammars", () => {
  it("parses a grammar", () => {
    expect(
      parseGrammarFromString_withClassic(`# This is a comment
   Abc:
     x0 \`string\`
     x1 \`string\`+
     x2 \`string\`*
     x3 \`string\`?
     y0 @Xyz
     y1 @Xyz*
     y2 @Xyz+
     y3 @Xyz?
     y4 @Xyz?
     z0 Pqr
     z1 Pqr+
     z2 Pqr*
     z3 Pqr?
  
   @Xyz:
     | Pqr
     | Stu
  
   Pqr:
     p \`number\`?
  
   # This is a comment
   Stu:
     s \`number\`
   `),
    ).toMatchSnapshot();
  });
});

describe("checking document validity", () => {
  it("thrown where there is a duplicate def (node)", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo {}
      Foo {}
   `),
    ).toThrow("Duplicate definition of 'Foo'");
  });

  it("thrown where there is a duplicate union def (node)", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo {}
      Bar {}
      @Bar = Foo
      @Bar = Bar
   `),
    ).toThrow("Duplicate definition of 'Bar'");
  });

  it("thrown where there is a duplicate def (node + union)", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo {}
      Bar {}
      @Foo = Bar
   `),
    ).toThrow("Duplicate definition of 'Foo'");
  });

  it("thrown when there is an unknown reference (node)", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo { bar: Ba }
      #          ^^ Typo, should fail
      Bar { }
   `),
    ).toThrow("Cannot find 'Ba'");
  });

  it("thrown when there is an unknown reference (node)", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo { bar: @Ba }
      #          ^^^ Typo, should fail
      @Bar = Qux | Mutt
      Qux {}
      Mutt {}
   `),
    ).toThrow("Cannot find '@Ba'");
  });

  it("thrown when there is an unused node def", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo { }
      Bar { }
   `),
    ).toThrow("Unused definition 'Bar'");
  });

  it("thrown when there is an unused union def", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo { }
      @Bar = Foo
   `),
    ).toThrow("Unused definition '@Bar'");
  });

  it.skip("[todo later] thrown when there is a circular ref", () => {
    expect(() =>
      parseGrammarFromString_withOhm(`
      Foo { foo: Foo }
   `),
    ).toThrow("Circular ref 'Foo'");
  });
});
