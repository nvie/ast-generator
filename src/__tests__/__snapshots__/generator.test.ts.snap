// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parsing grammars > parses a grammar 1`] = `
{
  "nodes": [
    {
      "fields": [
        {
          "name": "x0",
          "pattern": {
            "name": "string",
            "ref": "Raw",
          },
        },
        {
          "name": "x1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        {
          "name": "x2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        {
          "name": "x3",
          "pattern": {
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "y0",
          "pattern": {
            "name": "Xyz",
            "ref": "NodeUnion",
          },
        },
        {
          "name": "y1",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        {
          "name": "y2",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        {
          "name": "y3",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "y4",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "z0",
          "pattern": {
            "name": "Pqr",
            "ref": "Node",
          },
        },
        {
          "name": "z1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        {
          "name": "z2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        {
          "name": "z3",
          "pattern": {
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "Optional",
          },
        },
      ],
      "fieldsByName": {
        "x0": {
          "name": "x0",
          "pattern": {
            "name": "string",
            "ref": "Raw",
          },
        },
        "x1": {
          "name": "x1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        "x2": {
          "name": "x2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        "x3": {
          "name": "x3",
          "pattern": {
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
        "y0": {
          "name": "y0",
          "pattern": {
            "name": "Xyz",
            "ref": "NodeUnion",
          },
        },
        "y1": {
          "name": "y1",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        "y2": {
          "name": "y2",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        "y3": {
          "name": "y3",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        "y4": {
          "name": "y4",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        "z0": {
          "name": "z0",
          "pattern": {
            "name": "Pqr",
            "ref": "Node",
          },
        },
        "z1": {
          "name": "z1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        "z2": {
          "name": "z2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        "z3": {
          "name": "z3",
          "pattern": {
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "Optional",
          },
        },
      },
      "name": "Abc",
    },
    {
      "fields": [
        {
          "name": "p",
          "pattern": {
            "of": {
              "name": "number",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
      ],
      "fieldsByName": {
        "p": {
          "name": "p",
          "pattern": {
            "of": {
              "name": "number",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
      },
      "name": "Pqr",
    },
    {
      "fields": [
        {
          "name": "s",
          "pattern": {
            "name": "number",
            "ref": "Raw",
          },
        },
      ],
      "fieldsByName": {
        "s": {
          "name": "s",
          "pattern": {
            "name": "number",
            "ref": "Raw",
          },
        },
      },
      "name": "Stu",
    },
  ],
  "nodesByName": {
    "Abc": {
      "fields": [
        {
          "name": "x0",
          "pattern": {
            "name": "string",
            "ref": "Raw",
          },
        },
        {
          "name": "x1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        {
          "name": "x2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        {
          "name": "x3",
          "pattern": {
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "y0",
          "pattern": {
            "name": "Xyz",
            "ref": "NodeUnion",
          },
        },
        {
          "name": "y1",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        {
          "name": "y2",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        {
          "name": "y3",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "y4",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        {
          "name": "z0",
          "pattern": {
            "name": "Pqr",
            "ref": "Node",
          },
        },
        {
          "name": "z1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        {
          "name": "z2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        {
          "name": "z3",
          "pattern": {
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "Optional",
          },
        },
      ],
      "fieldsByName": {
        "x0": {
          "name": "x0",
          "pattern": {
            "name": "string",
            "ref": "Raw",
          },
        },
        "x1": {
          "name": "x1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        "x2": {
          "name": "x2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "List",
          },
        },
        "x3": {
          "name": "x3",
          "pattern": {
            "of": {
              "name": "string",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
        "y0": {
          "name": "y0",
          "pattern": {
            "name": "Xyz",
            "ref": "NodeUnion",
          },
        },
        "y1": {
          "name": "y1",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        "y2": {
          "name": "y2",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "List",
          },
        },
        "y3": {
          "name": "y3",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        "y4": {
          "name": "y4",
          "pattern": {
            "of": {
              "name": "Xyz",
              "ref": "NodeUnion",
            },
            "ref": "Optional",
          },
        },
        "z0": {
          "name": "z0",
          "pattern": {
            "name": "Pqr",
            "ref": "Node",
          },
        },
        "z1": {
          "name": "z1",
          "pattern": {
            "min": 1,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        "z2": {
          "name": "z2",
          "pattern": {
            "min": 0,
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "List",
          },
        },
        "z3": {
          "name": "z3",
          "pattern": {
            "of": {
              "name": "Pqr",
              "ref": "Node",
            },
            "ref": "Optional",
          },
        },
      },
      "name": "Abc",
    },
    "Pqr": {
      "fields": [
        {
          "name": "p",
          "pattern": {
            "of": {
              "name": "number",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
      ],
      "fieldsByName": {
        "p": {
          "name": "p",
          "pattern": {
            "of": {
              "name": "number",
              "ref": "Raw",
            },
            "ref": "Optional",
          },
        },
      },
      "name": "Pqr",
    },
    "Stu": {
      "fields": [
        {
          "name": "s",
          "pattern": {
            "name": "number",
            "ref": "Raw",
          },
        },
      ],
      "fieldsByName": {
        "s": {
          "name": "s",
          "pattern": {
            "name": "number",
            "ref": "Raw",
          },
        },
      },
      "name": "Stu",
    },
  },
  "startNode": "Abc",
  "unions": [
    {
      "members": [
        {
          "name": "Pqr",
          "ref": "Node",
        },
        {
          "name": "Stu",
          "ref": "Node",
        },
      ],
      "name": "Xyz",
    },
  ],
  "unionsByName": {
    "Xyz": {
      "members": [
        {
          "name": "Pqr",
          "ref": "Node",
        },
        {
          "name": "Stu",
          "ref": "Node",
        },
      ],
      "name": "Xyz",
    },
  },
}
`;
